napari/layers/image/experimental/octree.py

```diff
@@ -0,0 +1,307 @@
+    pos : Tuple[float, float]
+        The x, y coordinates of the chunk.
+    size : float
+        The size of the chunk, the chunk is square/cubic.
+    """
+
+    def __init__(self, data: ArrayLike, pos: Tuple[float, float], size: float):
+        self.data = data
+        self.pos = pos
+        self.size = size
+
+
+def _create_tiles(array: np.ndarray, tile_size: int) -> np.ndarray:
+    """
+    Return an NxM array of (tile_size, tile_size) ndarrays except the edge
+    tiles might be smaller if the array did not divide evenly.
+
+    TODO_OCTREE: slices_from_chunks from dask.array.core possibly does
+    the same thing, if we are going to use this in production.
+
+    Parameters
+    ----------
+    array : np.ndarray
+        The array to create tiles out of.
+    tiles_size : int
+        Edge length of the square tiles.
+    """
+    if array.ndim != 3:
+        raise ValueError(f"Unexpected array dimension {array.ndim}")
+    (rows, cols, _) = array.shape
+
+    tiles = []
+
+    row = 0
+    while row < rows:
+        row_tiles = []
+        col = 0
+        while col < cols:
+            tile = array[row : row + tile_size, col : col + tile_size, :]
+            row_tiles.append(tile)
+            col += tile_size
+        tiles.append(row_tiles)
+        row += tile_size
+
+    return tiles
+
+
+def _create_downsampled_tile(
+    ul: np.ndarray, ur: np.ndarray, ll: np.ndarray, lr: np.ndarray
+) -> np.ndarray:
+    """Create one parent tile from four child tiles.
+
+    Parameters
+    ----------
+    ul : np.ndarray
+        Upper left child tile.
+    ur : np.ndarray
+        Upper right child tile.
+    ll : np.ndarray
+        Lower left child tile.
+    lr : np.ndarray
+        Lower right child tile.
+    """
+    row1 = np.hstack((ul, ur))
+    row2 = np.hstack((ll, lr))
+    combined_tile = np.vstack((row1, row2))
+
+    # Down sample by half.
+    return ndi.zoom(combined_tile, [0.5, 0.5, 1])
+
+
+def _create_coarser_level(tiles: TileArray) -> TileArray:
+    """Return the next coarser level of tiles.
+
+    Combine each 2x2 group of tiles into one downsampled tile.
+
+    tiles : TileArray
+        The tiles to combine.
+    """
+
+    new_tiles = []
+
+    for row in range(0, len(tiles), 2):
+        row_tiles = []
+        for col in range(0, len(tiles[row]), 2):
+            tile = _create_downsampled_tile(
+                tiles[row][col],
+                tiles[row][col + 1],
+                tiles[row + 1][col],
+                tiles[row + 1][col + 1],
+            )
+            row_tiles.append(tile)
+        new_tiles.append(row_tiles)
+
+    return new_tiles
+
+
+def _print_level_tiles(tiles):
+    """Print information about these tiles.
+    """
+    num_rows = len(tiles)
+
+    num_cols = None
+    for row in tiles:
+        if num_cols is None:
+            num_cols = len(row)
+        else:
+            assert num_cols == len(row)
+
+    print(f"{num_rows} x {num_cols} = {num_rows * num_cols}")
```

Comment:
is there a reason in this file you're using all the print statements instead of the logging ones? Doesn't need to be fixed now, but maybe something for the future

Reply:
No real reason. For super temporary stuff I still sometimes use prints. But migrate things to logging which I feel like we'll always want available. Always want to be able to toggle back on.

Reply:
As we discussed once, there should be no prints in mainstream code. That could be CI enforced someday. Although every project I've worked on people just shout "your stuff is printing junk!" and someone gets rid of it. Has not been a huge issue. Yet stray prints pop up fairly regularly.

Reply:
sounds good, we can address later. i like the idea of a CI option at some point